src/emulate.c

```diff
@@ -1356,6 +1410,78 @@ static void block_translate(riscv_t *rv, block_t *block)
+    }                                                                      \
+    if (count > 1) {                                                       \
+        ir->opcode = IIF(RW)(rv_insn_fuse4, rv_insn_fuse3);                \
+        ir->fuse = malloc(count * sizeof(opcode_fuse_t));                  \
+        ir->imm2 = count;                                                  \
+        memcpy(ir->fuse, ir, sizeof(opcode_fuse_t));                       \
+        ir->impl = dispatch_table[ir->opcode];                             \
+        for (int j = 1; j < count; j++) {                                  \
+            next_ir = ir + j;                                              \
+            memcpy(ir->fuse + j, next_ir, sizeof(opcode_fuse_t));          \
+            next_ir->opcode = rv_insn_empty;                               \
+            next_ir->impl = dispatch_table[next_ir->opcode];               \
+        }                                                                  \
+    }                                                                      \
+    break;
+
+
+/* examine whether instructions in a block match a specific pattern. If so,
+ * rewrite them into fused instructions. */
+static void match_pattern(block_t *block)
+{
+    for (uint32_t i = 0; i < block->n_insn - 1; i++) {
+        rv_insn_t *ir = block->ir + i, *next_ir = NULL;
+        int32_t count = 0, sign = 1;
+        switch (ir->opcode) {
+        case rv_insn_auipc:
+            next_ir = ir + 1;
+            if (next_ir->opcode == rv_insn_addi && ir->rd == next_ir->rs1) {
+                /* the destination register of instruction auipc is equal to the
+                 * source register 1 of next instruction addi */
+                ir->opcode = rv_insn_fuse1;
+                ir->rd = next_ir->rd;
+                ir->imm2 = next_ir->imm;
+                ir->impl = dispatch_table[ir->opcode];
+                next_ir->opcode = rv_insn_empty;
+                next_ir->impl = dispatch_table[next_ir->opcode];
+            } else if (next_ir->opcode == rv_insn_add &&
+                       ir->rd == next_ir->rs2) {
+                /* the destination register of instruction auipc is equal to the
+                 * source register 2 of next instruction add */
+                ir->opcode = rv_insn_fuse2;
+                ir->rd = next_ir->rd;
+                ir->rs1 = next_ir->rs1;
+                ir->impl = dispatch_table[ir->opcode];
+                next_ir->opcode = rv_insn_empty;
+                next_ir->impl = dispatch_table[next_ir->opcode];
+            }
+            break;
+        case rv_insn_sw:
+            /* If the memory addresses of a sequence of store instructions for
+             * data are contiguous, pack these instructions. */
+            pack_memory_operation(0);
+        case rv_insn_lw:
+            /* If the memory addresses of a sequence of load instructions for
```

Comment:
Remove this comment. Write comments prior to `case rv_insn_sw` and `case rv_insn_lw`.